File: go/msg.go
/* Copyright © 2026 Voxgig Ltd, MIT License. */

package model

import (
	"errors"
	"sort"
	"strconv"
	"strings"
)

func MsgProducer(b *Build, ctx *BuildContext) ProducerResult {
	pr := ProducerResult{OK: true, Name: "msg", Step: ctx.Step, Active: true}

	problems := checkMsg(b.Model)
	if len(problems) == 0 {
		return pr
	}

	pr.OK = false
	for _, problem := range problems {
		pr.Errs = append(pr.Errs, errors.New(problem))
	}

	// Unlike the TypeScript producer, this does not add the errors to the
	// build itself: runProducer merges a failed producer's Errs. Both
	// implementations end up with the same errors on the build.
	b.Log.Error("msg-invalid", nil, strings.Join(problems, "; "))

	return pr
}

// checkMsg validates the message declarations in main.msg, returning one
// message per problem found (none when the model is valid, which includes a
// model with no messages at all, or only a legacy chain).
func checkMsg(model map[string]any) []string {
	main := asMap(model["main"])
	if main == nil {
		return nil
	}

	if list, isList := main["msg"].([]any); isList {
		return checkMsgList(list)
	}

	if chain := asMap(main["msg"]); chain != nil {
		return checkMsgChain(chain)
	}

	return nil
}

// checkMsgList validates the declared shape: a list of definitions.
func checkMsgList(msg []any) []string {
	var problems []string

	seen := map[string]int{}

	for mI, elem := range msg {
		def := asMap(elem)
		if def == nil {
			problems = append(problems, msgerr(mI, "is not a message definition"))
			continue
		}

		pat, isList := def["pat"].([]any)
		if !isList {
			problems = append(problems, msgerr(mI, "has no pat list"))
			continue
		}

		if len(pat) == 0 {
			problems = append(problems, msgerr(mI, "pat declares no pattern pairs"))
			continue
		}

		pairs := make([]string, 0, len(pat))
		canon := make([]string, 0, len(pat))
		wellFormed := true

		for pI, pelem := range pat {
			pair := asMap(pelem)
			if len(pair) != 1 {
				problems = append(problems, msgerr(mI,
					"pat pair "+strconv.Itoa(pI)+" is not a single key:value pair"))
				wellFormed = false
				break
			}

			var key string
			for k := range pair {
				key = k
			}

			val, isStr := pair[key].(string)
			if !isStr {
				problems = append(problems, msgerr(mI,
					"pat pair "+strconv.Itoa(pI)+" ("+key+") value is not a string"))
				wellFormed = false
				break
			}

			pairs = append(pairs, key+":"+val)
			canon = append(canon, strconv.Quote(key)+":"+strconv.Quote(val))
		}

		if !wellFormed {
			continue
		}

		// file names the action file, overriding the last-pattern-pair
		// convention. A non-string would reach the consumers as one.
		if file, has := def["file"]; has {
			if _, isStr := file.(string); !isStr {
				problems = append(problems, msgerr(mI, "file is not a string"))
			}
		}

		canonKey := strings.Join(canon, ",")
		if first, dup := seen[canonKey]; dup {
			problems = append(problems, msgerr(mI,
				"pat ["+strings.Join(pairs, ",")+"] is already declared by msg ["+
					strconv.Itoa(first)+"]"))
		} else {
			seen[canonKey] = mI
		}
	}

	return problems
}

func checkMsgChain(msg map[string]any) []string {
	var problems []string

	names := make([]string, 0, len(msg))
	for name := range msg {
		names = append(names, name)
	}
	// Byte order. Go map iteration is random, so without this the problems
	// would come out in a different order on every run - and in a different
	// order from the TypeScript producer, which sorts its keys the same way.
	sort.Strings(names)

	for _, name := range names {
		if isMsgDef(msg[name]) {
			problems = append(problems, `model msg "`+name+`": a message `+
				`definition must be declared in the main.msg list, not as a `+
				`keyed entry (main: msg: [ { pat: [...] } ])`)
		}
	}

	return problems
}

// isMsgDef reports whether a value is a message definition: a map declaring
// its pattern as a list.
func isMsgDef(val any) bool {
	def := asMap(val)
	if def == nil {
		return false
	}
	_, isList := def["pat"].([]any)
	return isList
}

func msgerr(index int, why string) string {
	return "model msg [" + strconv.Itoa(index) + "]: " + why
}

File: ts/src/producer/msg.ts
/* Copyright © 2026 Voxgig Ltd, MIT License. */

import type { Build, Producer, BuildContext, ProducerResult } from '../types'




// Report a problem against the definition it belongs to.
function msgerr(index: number, why: string): string {
  return 'model msg [' + index + ']: ' + why
}


function isObj(val: any): boolean {
  return null != val && 'object' === typeof val && !Array.isArray(val)
}


// A message definition declares its pattern as a list.
function isMsgDef(val: any): boolean {
  return isObj(val) && Array.isArray(val.pat)
}


function sortNames(names: string[]): string[] {
  return names.sort((a, b) => Buffer.compare(Buffer.from(a), Buffer.from(b)))
}


// Validate the message declarations in main.msg, returning one message per
// problem found (empty when the model is valid, which includes a model with
// no messages at all, or only a legacy chain).
function checkMsg(model: any): string[] {
  const msg = model?.main?.msg

  if (Array.isArray(msg)) {
    return checkMsgList(msg)
  }

  if (isObj(msg)) {
    return checkMsgChain(msg)
  }

  return []
}


// The declared shape: a list of definitions.
function checkMsgList(msg: any[]): string[] {
  const problems: string[] = []

  // Canonical pattern -> the index that declared it first.
  const seen: { [canon: string]: number } = {}

  for (let mI = 0; mI < msg.length; mI++) {
    const def = msg[mI]

    if (!isObj(def)) {
      problems.push(msgerr(mI, 'is not a message definition'))
      continue
    }

    if (!Array.isArray(def.pat)) {
      problems.push(msgerr(mI, 'has no pat list'))
      continue
    }

    const pat: any[] = def.pat

    if (0 === pat.length) {
      problems.push(msgerr(mI, 'pat declares no pattern pairs'))
      continue
    }

    const pairs: string[] = []
    const canon: string[] = []
    let wellFormed = true

    for (let pI = 0; pI < pat.length; pI++) {
      const pair = pat[pI]
      const keys = isObj(pair) ? Object.keys(pair) : []

      if (1 !== keys.length) {
        problems.push(msgerr(mI, 'pat pair ' + pI +
          ' is not a single key:value pair'))
        wellFormed = false
        break
      }

      const key = keys[0]
      const val = pair[key]

      if ('string' !== typeof val) {
        problems.push(msgerr(mI, 'pat pair ' + pI + ' (' + key +
          ') value is not a string'))
        wellFormed = false
        break
      }

      pairs.push(key + ':' + val)
      canon.push(JSON.stringify(key) + ':' + JSON.stringify(val))
    }

    if (!wellFormed) {
      continue
    }

    // `file` names the action file, overriding the last-pattern-pair
    // convention. A non-string would reach the consumers as one.
    if (undefined !== def.file && 'string' !== typeof def.file) {
      problems.push(msgerr(mI, 'file is not a string'))
    }

    const canonKey = canon.join(',')
    if (undefined === seen[canonKey]) {
      seen[canonKey] = mI
    }
    else {
      problems.push(msgerr(mI, 'pat [' + pairs.join(',') +
        '] is already declared by msg [' + seen[canonKey] + ']'))
    }
  }

  return problems
}


function checkMsgChain(msg: any): string[] {
  const problems: string[] = []

  for (const name of sortNames(Object.keys(msg))) {
    if (isMsgDef(msg[name])) {
      problems.push('model msg "' + name + '": a message definition must be ' +
        'declared in the main.msg list, not as a keyed entry ' +
        '(main: msg: [ { pat: [...] } ])')
    }
  }

  return problems
}


const msg_producer: Producer = async (build: Build, ctx: BuildContext) => {
  const pr: ProducerResult = {
    ok: true,
    name: 'msg',
    reload: false,
    step: ctx.step,
    active: true,
    errs: [],
    runlog: []
  }

  const problems = checkMsg(build.model)

  if (0 < problems.length) {
    pr.ok = false
    pr.errs = problems.map((problem) => new Error(problem))

    // Add them to the build too. BuildImpl.run collects the errors a producer
    // THROWS, but not the ones it returns, so a returned error would
    // otherwise be missing from the BuildResult. (The Go port needs no such
    // push: its runProducer merges a failed producer's Errs itself. Both
    // implementations end up with the same errors on the build.)
    build.errs.push(...pr.errs)

    build.log.error({
      point: 'msg-invalid',
      count: problems.length,
      note: problems.join('; ')
    })
  }

  return pr
}


export {
  msg_producer,
  checkMsg,
}
